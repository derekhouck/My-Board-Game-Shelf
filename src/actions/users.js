import { SubmissionError } from 'redux-form';

import { API_BASE_URL } from '../config';
import { normalizeResponseErrors, startLoading, stopLoading } from './utils';
import { clearAuth } from './auth'
import { clearAuthToken } from '../local-storage';

export const ADD_GAME = 'ADD_GAME';
export const addGame = game => ({
  type: ADD_GAME,
  game
});

export const FETCH_USER_GAMES_SUCCESS = 'FETCH_USER_GAMES_SUCCESS';
export const fetchUserGamesSuccess = games => ({
  type: FETCH_USER_GAMES_SUCCESS,
  games
});

export const FETCH_USERS_REQUEST = 'FETCH_USERS_REQUEST';
export const fetchUsersRequest = () => ({
  type: FETCH_USERS_REQUEST
});

export const FETCH_USERS_SUCCESS = 'FETCH_USERS_SUCCESS';
export const fetchUsersSuccess = users => ({
  type: FETCH_USERS_SUCCESS,
  users
});

export const REMOVE_USER_GAME = 'REMOVE_USER_GAME';
export const removeUserGame = game => ({
  type: REMOVE_USER_GAME,
  game
});

export const SET_EDITING = 'SET_EDITING';
export const setEditing = editing => ({
  type: SET_EDITING,
  editing
});

export const TOGGLE_DELETING = 'TOGGLE_DELETING';
export const toggleDeleting = () => ({
  type: TOGGLE_DELETING
});

export const USERS_ERROR = 'USERS_ERROR';
export const usersError = error => ({
  type: USERS_ERROR,
  error
});

export const addGameToShelf = game => (dispatch, getState) => {
  const { authToken, currentUser } = getState().auth;

  dispatch(startLoading());
  return fetch(`${API_BASE_URL}/users/${currentUser.id}/games`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${authToken}`,
      'content-type': 'application/json'
    },
    body: JSON.stringify({ gameId: game.id })
  })
    .then(res => normalizeResponseErrors(res))
    .then(() => {
      dispatch(addGame(game));
      dispatch(stopLoading());
    })
    .catch(err => {
      dispatch(usersError(err));
      dispatch(stopLoading());
    });
};

export const editUser = user => (dispatch, getState) => {
  const authToken = getState().auth.authToken;

  return fetch(`${API_BASE_URL}/users/${user.id}`, {
    method: 'PUT',
    headers: {
      Authorization: `Bearer ${authToken}`,
      'content-type': 'application/json'
    },
    body: JSON.stringify(user)
  })
    .then(res => normalizeResponseErrors(res))
    .then(res => res.json())
    .catch(err => {
      const { reason, message, location } = err;
      if (reason === 'ValidationError') {
        // Convert ValidationErrors into SubmissionErrors for Redux Form
        return Promise.reject(
          new SubmissionError({
            [location]: message
          })
        );
      }
    });
};

export const fetchUsers = () => (dispatch, getState) => {
  const authToken = getState().auth.authToken;
  dispatch(fetchUsersRequest());
  return fetch(`${API_BASE_URL}/users`, {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${authToken}`
    }
  })
    .then(res => normalizeResponseErrors(res))
    .then(res => res.json())
    .then(users => dispatch(fetchUsersSuccess(users)))
    .catch(err => {
      dispatch(usersError(err));
    });
}

export const fetchUserGames = (userId, filters) => (dispatch, getState) => {
  let url = new URL(`${API_BASE_URL}/users/${userId}/games`);
  if (filters) {
    Object.keys(filters).forEach(key => {
      return url.searchParams.append(key, filters[key])
    });
  }
  const authToken = getState().auth.authToken;

  dispatch(startLoading());

  return fetch(url, {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${authToken}`
    }
  })
    .then(res => normalizeResponseErrors(res))
    .then(res => res.json())
    .then(games => {
      dispatch(fetchUserGamesSuccess(games));
      dispatch(stopLoading());
    })
    .catch(err => {
      dispatch(usersError(err));
      dispatch(stopLoading());
    });
}

export const removeGameFromShelf = game => (dispatch, getState) => {
  const { authToken, currentUser } = getState().auth;

  dispatch(startLoading());

  return fetch(`${API_BASE_URL}/users/${currentUser.id}/games`, {
    method: 'DELETE',
    headers: {
      Authorization: `Bearer ${authToken}`,
      'content-type': 'application/json'
    },
    body: JSON.stringify({ gameId: game.id })
  })
    .then(res => normalizeResponseErrors(res))
    .then(() => {
      dispatch(removeUserGame(game));
      dispatch(stopLoading());
    })
    .catch(err => {
      dispatch(usersError(err));
      dispatch(stopLoading());
    });
}

export const registerUser = user => dispatch => {
  return fetch(`${API_BASE_URL}/users`, {
    method: 'POST',
    headers: {
      'content-type': 'application/json'
    },
    body: JSON.stringify(user)
  })
    .then(res => normalizeResponseErrors(res))
    .then(res => res.json())
    .catch(err => {
      const { reason, message, location } = err;
      if (reason === 'ValidationError') {
        // Convert ValidationErrors into SubmissionErrors for Redux Form
        return Promise.reject(
          new SubmissionError({
            [location]: message
          })
        );
      }
    });
};

export const deleteUser = user => (dispatch, getState) => {
  const authToken = getState().auth.authToken;
  return fetch(`${API_BASE_URL}/users/${user.id}`, {
    method: 'DELETE',
    headers: {
      Authorization: `Bearer ${authToken}`
    }
  })
    .then(() => {
      dispatch(clearAuth());
      clearAuthToken();
    })
    .catch(err => {
      const { reason, message, location } = err;
      if (reason === 'ValidationError') {
        // Convert ValidationErrors into SubmissionErrors for Redux Form
        return Promise.reject(
          new SubmissionError({
            [location]: message
          })
        );
      }
    });
}