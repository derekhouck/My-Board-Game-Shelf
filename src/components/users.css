.users {
  border: 1px solid #19000b;
  border-collapse: collapse;
  text-align: left;
}

.users td, .users th {
  padding: 0.5em;
  border-left: 1px solid #19000b;
  border-right: 1px solid #19000b;
}

.users td {
}

.users th {
  background-color: #4f2119;
  color: #f9fafa;
}

.users tr:nth-child(even) {
  background-color: hsla(18, 70%, 61%, 0.5);
}

.users tr:active, .users tr:focus, .users tr:hover {
  background-color: #e18258;
}