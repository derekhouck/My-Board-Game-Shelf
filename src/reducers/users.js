import {
  FETCH_USER_GAMES_SUCCESS,
  FETCH_USERS_REQUEST,
  FETCH_USERS_SUCCESS,
  REMOVE_GAME,
  TOGGLE_DELETING,
  USERS_ERROR
} from '../actions/users';

const initialState = {
  games: [],
  users: [],
  loading: false,
  deleting: false,
  error: null
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case FETCH_USER_GAMES_SUCCESS:
      return Object.assign({}, state, {
        games: action.games
      });
    case FETCH_USERS_REQUEST:
      return Object.assign({}, state, {
        loading: true,
        error: null
      });
    case FETCH_USERS_SUCCESS:
      return Object.assign({}, state, {
        loading: false,
        users: action.users
      });
    case REMOVE_GAME:
      return Object.assign({}, state, {
        games: state.games.filter(game => game.id !== action.game.id)
      });
    case TOGGLE_DELETING:
      return Object.assign({}, state, {
        deleting: !state.deleting
      });
    case USERS_ERROR:
      return Object.assign({}, state, {
        loading: false,
        error: action.error
      });
    default:
      return state;
  }
}