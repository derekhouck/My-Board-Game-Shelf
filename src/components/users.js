import React from "react";
import requiresLogin from "./requires-login";
import { connect } from "react-redux";
import { Redirect } from 'react-router-dom';
import { fetchUsers } from "../actions/users";
import './users.css';

export class Users extends React.Component {
  componentDidMount() {
    const { dispatch, isAdmin } = this.props;
    if (isAdmin) {
      dispatch(fetchUsers());
    }
  }

  renderUsers() {
    let body;

    if (this.props.error) {
      body = (
        <div className="message message-error">{this.props.error.message}</div>
      );
    } else if (this.props.loading) {
      body = <div className="message message-default">Loading users...</div>;
    } else {
      const users = this.props.users.map(user => (
        <tr className="user" key={user.id} id={user.id}>
          <td>
            {user.username}
          </td>
          <td>
            {user.name}
          </td>
          <td>
            {user.admin ? "Yes" : "No"}
          </td>
        </tr>
      ));

      body = (
        <table className="users">
          <tr>
            <th>Username</th>
            <th>Name</th>
            <th>Admin</th>
          </tr>
          {users}
        </table>
      );
    }

    return body;
  }

  render() {
    const { isAdmin } = this.props;
    return isAdmin
      ? <section>
        <h2>Users</h2>
        {this.renderUsers()}
      </section>
      : <Redirect to="/" />;
  }
}

const mapStateToProps = state => ({
  error: state.users.error,
  isAdmin: state.auth.currentUser.admin,
  users: state.users.users,
  loading: state.users.loading,
});

export default requiresLogin()(connect(mapStateToProps)(Users));
