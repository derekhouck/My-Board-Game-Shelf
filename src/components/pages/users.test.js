import React from 'react';
import { shallow } from 'enzyme';
import { Users } from './users';

describe('<Users />', function () {
  it('renders without crashing', () => {
    const callback = jest.fn();
    const wrapper = shallow(<Users users={[]} dispatch={callback} />);
    expect(wrapper).toMatchSnapshot();
  });
});
