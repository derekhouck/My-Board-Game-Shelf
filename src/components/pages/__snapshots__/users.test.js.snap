// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Users /> renders without crashing 1`] = `ShallowWrapper {}`;
